simphony/cuds/model.py

```diff
@@ -39,119 +55,179 @@ def __init__(self):
         # that will return the corresponding value.
         self._map = {}
 
-    @staticmethod
-    def _is_cuds_component(obj):
-        """Check whether the object is a CUDS component."""
-        # When I see a bird that walks like a duck and swims like a duck
-        # and quacks like a duck, I call that bird a duck
-        # TODO: replace with metadata type checks
-        return all([hasattr(obj, 'name'),
-                    hasattr(obj, 'data')])
+        # Another map to keep a mapping between names and ids
+        self._name_to_id_map = {}
+
+    @property
+    def uid(self):
+        return self._uid
+
+    @uid.setter
+    def uid(self, value):
+        raise TypeError("uid is readonly.")
+
+    @uid.deleter
+    def uid(self):
+        raise TypeError("Can't delete uid.")
 
     @staticmethod
     def _is_dataset(obj):
         """Check if the object is a dataset."""
-        return isinstance(obj, (ABCParticles, ABCLattice, ABCMesh))
+        return isinstance(obj, (ABCParticles,
+                                ABCLattice,
+                                ABCMesh,
+                                api.DataSet))
 
     @property
     def data(self):
         """Data container for CUDS items."""
         return DataContainer(self._data)
```

Comment:
What is the reasoning for creating a (shallow) copy of the `self._data`. If the purpose is to have a snapshot then this will work only if the contained `keys` and `values` are immutable. In any other case we have an object that references internal information that can change without anybody knowing about them.

If I am correct only the engine functions and returned dataset are required support the snapshot principle. Does this extent to simulation? what is there a reason for this?